### DISCUSSION TITLE
[Upload Model] Secure model location URL

### DISCUSSION THREAD

**Discussion by khoaisohd** [opensearch-project/ml-commons#1534] [General]
**Background**
Hi experts, today customers are able to upload models that already exist on another server, such as GitHub or S3 by providing the model URL. In the implementation, it looks like the model URL needs to be public so that it can be downloadable. 
https://opensearch.org/docs/latest/ml-commons-plugin/ml-framework/#upload-model-to-opensearch
**Current use case**
In our use case, our customers want to bring their own model but they don't want to make it publicly, for example, the model is stored in a private S3 bucket
**Question**
Do we have a plan to support private access for model location URL, or is there any approach to make the model location access secure?

**Comment by austintlee** [comment]
You mean ml-commons does not use the AWS SDK for S3 to download models hosted in an S3 bucket?

**Comment by austintlee** [comment]
Is this covered by #1371 ?

**Reply by khoaisohd** [reply]
Yeah it is covered by https://github.com/opensearch-project/ml-commons/issues/1371

**Comment by khoaisohd** [comment]
Hi @austintlee , yeah ml-commons does not use the AWS SDK for S3 to download models hosted in an S3 bucket.
I'm curious how people work with models in private S3 bucket, does people get pre-signed URL (within a short period of time) and then use that link to upload model?

**Reply by austintlee** [reply]
I am not sure, but I would guess people just download to a local storage and upload from there?

**Reply by ylwu-amzn** [reply]
@khoaisohd , your understanding is correct. ml-commons only supports public URL, so pre-signed URL works.

The second way is using pretrained model https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/

@austintlee , correct , people can also download to a local storage , then upload from local file.

**Reply by khoaisohd** [reply]
I see, so here is what customers can do to use the model in a private bucket
* Create a pre-signed URL for the model in a private bucket with an expiration time of 5 minutes (assumed that should be enough to upload the model)
* Upload model using pre-signed URL

Do you think it is good enough or do we still need to wait for https://github.com/opensearch-project/ml-commons/issues/1371
I'm considering the tradeoff between complexity and security perspective, it looks like the model URL only needs to be public for a short period of time
